Cleanly handle NetCreate ITEST as well as 2018

jdanish opened the pull request:
These changes are required to switch from using netcreate-2018 to Net.Create 2.0 / netcreate-itest (https://github.com/netcreateorg/netcreate-itest). DO NOT switch to this version if you are not prepared to use the newest Net.Create.

The new Net.Create version now requires a higher node version (18.16.0) and involved moving several files around. Notably, we are no-longer using a build sub-folder, and the configuration files have shifted. Also, the default net.create install folder has been renamed to netcreate-itest from netcreate-2018.

Commit:
Changed file paths to reflect the new netcreate-itest and no build directory

Commit:
Updated to match netcreate

Commit:
Merge commit 'e4495f404c65696f299c4c402dc45252c7c5e42a' into dev-hd/tryingtoupdate

* commit 'e4495f404c65696f299c4c402dc45252c7c5e42a':
  Update .nvmrc to 18.16 for itest
  Update ReadMe.md for netcreate-itest.

Commit:
Adjusted to account for the new file location

daveseah committed:
add missing ignore settings to repo to prevent accidental repo corruption

(1) The repo did not have a .gitignore file, which excludes files/folders that should NEVER be committed to the repo.
(2) The repo uses subfolder GIT repos that can't directly be ignored in VSCODE, so .vscode/settings now ignores them.
These changes not only avoid accidental git commits by tired programmers, but improves developer "quality of life" by filtering-out irrelevant (and potentially confusing/dangerous) changes detected by git when these safeguards are not in place.

daveseah committed:
add missing editor configuration, linting, req packages

To ensure that there is consistency in style in all of our related codebases, nc-multiplex should also use the same linting
rules as the core `netcreate-2018` and `netcreate-itest` modules. Currently these files are most up-to-date in itest,
as 2018 had linting completely disabled which helped break code formatting consistency. The new itest rules are much more
forgiving now and deliver good results, so they have been added to this repo to fix the many inconsistencies that are in it.

daveseah committed:
add nvm version mismatch warning

Corrected the .nvmrc version format to match the output of `node --version` so the runtime warning for NODE version matches as expected.

daveseah committed:
add runtime repo path detection and adaptation

checks for existence of netcreate-itest or netcreate-2018 or both, returns appropriate warnings and strings.

daveseah committed:
Update `nc-multiplex` to use dynamic paths from `nc-launch-config`; other fixes

fix (1) rename `nc-start` to more functional name `nc-launch-instance`
fix (2) linter format to code standards

daveseah commented:
I reviewed the pull request and added some bullet-proofing so this can be merged into the main `dev` branch rather than having to maintain multiple "temporary branches" that are not documented. 

The main magic is now in the new module `nc-launch-config.js`, which scans for a list of repos that are in the `REPO_PATHS` data structure. It then constructs the necessary paths to be used by `nc-multiple.js` so it loads the right data.

Other opportunistic fixes:

* **linting** Added the default updated `.eslintrc` stuff for autoformatting, as this doesn't affect `netcreate-2018` and it's good to have enabled. It's already in `netcreate-itest`
* **node version warning** when mismatch is detected, as the node version has been updated to match that of `itest`. 
* **gitignore** and **subdir repo ignore** has been added to lower the possibility of screwing up a netcreate repo, as well as reduce noise from the editor

## TESTING

Install both `netcreate-2018` and `netcreate-itest`

* [ ] Run `node ./nc-multiplex` and note the new warning message, make sure it works. It defaults to `netcreate-2018`.
* [ ] Rename `netcreate-2018` to `nc-2018` and run it again. It should default to `netcreate-itest` and say so in the terminal console
* [ ] Rename `netcreate-itest` to `nc-itest` and run it again. You should see an error.

daveseah committed:
fix issue with hardcoded `SHELL.cd('netcreate-itest')`

and also prettify output for single matched repo

daveseah committed:
claimfix issue #4: Fix use of incorrect path NC_BUILD_PATH, replace with NC_SERVER_PATH

daveseah committed:
claimfix #3 multiplex now correctly intercepts the right path for `netcreate-config.js` to deliver correct config

The multiplexer overrides the default netcreate-config.js in the package by intercepting the path to it in the
public web folder; however, these paths are different in netcreate-2018 versus netcreate-itest, and therefore the
Express app.get() was NOT substituting the correct file. Instead, the packaged version was loaded normally.

daveseah commented:
Additional fixes to issue #3 and issue #4

benloh commented:
Confirmed that everything runs, including tokens.
Thanks for all the help!
